src/dmd/cond.d

```diff
@@ -149,15 +149,25 @@ extern (C++) final class StaticForeach : RootObject
         sc = sc.endCTFE();
         el = el.optimize(WANTvalue);
         el = el.ctfeInterpret();
+        Expressions *es;
         if (el.op == TOK.int64)
         {
-            dinteger_t length = el.toInteger();
-            auto es = new Expressions();
-            foreach (i; 0 .. length)
+            if (aggr.isArrayLiteralExp())
             {
-                auto index = new IntegerExp(loc, i, Type.tsize_t);
-                auto value = new IndexExp(aggr.loc, aggr, index);
-                es.push(value);
+                // Directly use the elements of the array created in lowerNonArrayAggregate()
+                auto ale = cast(ArrayLiteralExp)aggr;
+                es = ale.elements;
+            }
+            else
+            {
+                const length = cast(size_t)el.toInteger();
```

Comment:
Restore dinteger_t

Reply:
Why?
Array length is `size_t` so it needs that for 32bit platforms.
There are `cast(size_t)el.toInteger()` all over the place.

Reply:
Those are all potential bugs.  Think 32bit host targeting 64bit.

Bottom line is _never_ _ever_ use size_t, always use the standard compiler types for dealing with integers (which one day will be converted to a 128-bit type).

Reply:
IntegerExps are normalized when converted to a host type, so there is no problem toInteger() having wrong value on 32bit host.

Reply:
I need the cast for the Array constructor or compilation will fail on 32bit.
There are potential bugs but now it's the only way to do it.

Reply:
That is a problem with Array then.  It's max size should not be bound by 32 bits.

Reply:
Yeah, and that's another battle.
In fact I pushed the code first with `dinteger_t`, look at the tester [log](https://auto-tester.puremagic.com/show-run.ghtml?projectid=1&runid=4090801&isPull=true)

Reply:
I would make sure `el.toInteger <= int.max`
and then cast to `int`

Reply:
As I said, this pattern of getting `n = e.toInteger()` casting it to `size_t` and creating an Array!T `new Array!T(n)` is repeated in many places in the compiler.
Adding more, in this case the integer returned after ArrayLengthExp comes from things that are always size_t for example .dim and .len from Array or StringExp (see resolveArrayLength()).
Also the code before my changes even pushed an Array (size_t dim) over a foreach with upper bound in dinteger_t (ulong), so the right thing to do in the future is to change the Array implementation to dinteger_t/ulong length, then to patch all the compiler.

Reply:
@ibuclaw
> > I need the cast for the Array constructor or compilation will fail on 32bit.
> > There are potential bugs but now it's the only way to do it.
>
> That is a problem with Array then. It's max size should not be bound by 32 bits.

This is how D array's are implemented, I don't really agree with the implementation either, but I don't think having 64-bit length on 32-bit binary would be any better.